src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs

```diff
@@ -3253,19 +3227,14 @@ internal void FinishContinuations()
         {
             Debug.Assert(continuationObject != null);
 
-            TplEtwProvider etw = TplEtwProvider.Log;
-            bool tplEtwProviderLoggingEnabled = etw.IsEnabled();
```

Comment:
I know @benaadams and others have been fine-tuning the performance of this code. Caching of `Log.IsEnabled` in a local looks like the kind of change that would be done as part of the fine-tuning. Are we going to lose any cycles when the caching is removed?

Reply:
Caching it in local variable make it much harder for illinker to remove the code

Reply:
`.Log` is a static readonly based `EventSource` of the sealed type `TplEtwProvider`; and `IsEnabled()` is a non-virtual => `m_eventSourceEnabled` so it should inline happily. 

Main impact here seems to be 1 additional read vs having the local (as they are pairs); so it probably wouldn't be much.


Not used here, but `DiagnosticSource` is a more problematic type as its `.IsEnabled` is virtual so caching its result is helpful; although using its derived type `DiagnosticListener` instead is better as it has a fast non-virtual `.IsEnabled()`.

Removing `ThreadPool`'s caching of the Log states in the `Dispatch` loop might be more of an issue though (or any `.IsEnabled(string,..)`) vairant where its doing more checking; which here are guarded by the fast `.IsEnabled()` checks.

Reply:
Long story short, I'll defer to @stephentoub ;)

Reply:
The caching in a local was done years ago, before IsEnabled could be devirtualized, and it saved a static read in various places.  If you can confirm that it IsEnabled previously wasn't being devirtualized and now is being devirtualized and inlined, I'm happy for the change, as that'll have a net benefit. If it was already getting devirtualized/inlined, then we should run some perf tests just to confirm that there's no measurable negative impact, though I suspect there won't be.

Reply:
> `IsEnabled` could be devirtualized

IsEnable is not virtual. I do not think devirtualization has anything to do with this.

Reply:
You're right; I was thinking of DiagnosticSource.IsEnabled rather than EventSource.IsEnabled.

Reply:
In which case, I would like to see some validation done that this doesn't regress anything.  I don't expect it will measurable, but still...

Reply:
I'll undo the cleanup, the whole etw setup in Task looks very fishy to me